lib/Mojo/WebSocketProxy/Backend/ConsumerGroups.pm

```diff
@@ -0,0 +1,345 @@
+
+            return Future->done unless $c && $c->tx;
+
+            if ($error eq 'Timeout') {
+                $api_response = $c->wsp_error($msg_type, 'RequestTimeout', 'Request is timed out.');
+            } else {
+                $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+            }
+            $rpc_failure_cb->($c, undef, $req_storage) if $rpc_failure_cb;
+
+            $c->send({json => $api_response}, $req_storage);
+        })->retain;
+
+    return;
+}
+
+=head2 request
+
+Sends request to backend service. The method accepts single unnamed argument:
+
+=over 4
+
+=item * C<request_data> - an C<arrayref> containing data for the item which is going to be put into redis stream.
+
+=back
+
+Returns future.
+Which will be marked as done in case getting response from backend server.
+And it'll be marked as failed in case of request timeout or in case of error putting request to redis stream.
+
+=cut
+
+sub request {
+    my ($self, $request_data) = @_;
+
+    my $complete_future = $self->loop->new_future;
+
+    $self->wait_for_messages();
+
+    my $sent_future = $self->_send_request($request_data)->then(
+        sub {
+            my ($msg_id) = @_;
+
+            $self->pending_requests->{$msg_id} = $complete_future;
+
+            $complete_future->on_cancel(sub { delete $self->pending_requests->{$msg_id} });
+
+            return Future->done;
+        });
+
+    return Future->wait_any($self->loop->timeout_future(after => $self->timeout), Future->needs_all($complete_future, $sent_future),);
+}
+
+sub _send_request {
+    my ($self, $request_data) = @_;
+
+    my $f = $self->loop->new_future;
+    $self->redis->_execute(
+        xadd => XADD => ('rpc_requests', '*', $request_data->@*),
+        sub {
+            my ($redis, $err, $msg_id) = @_;
+
+            return $f->fail($err) if $err;
+
+            return $f->done($msg_id);
+        });
+
+    return $f;
+}
+
+=head2 wait_for_messages
+
+By using redis subscription, we subscribe on channel for receiving responses from backend server.
+We'll use uniq id generated by L<whoami> as subscription channel.
+Subscription will be done only once within first request to backend server.
+
+=cut
+
+sub wait_for_messages {
+    my ($self) = @_;
+    $self->{already_waiting} //= $self->redis->subscribe(
+        [$self->whoami],
+        $self->$curry::weak(
+            sub {
+                my ($self) = @_;
+                $self->redis->on('message', $self->$curry::weak('_on_message'));
+            }));
+
+    return;
+}
+
+sub _on_message {
+    my ($self, $redis, $raw_message) = @_;
+
+    my $message = eval { decode_json_utf8($raw_message) };
+
+    return unless ref $message eq 'HASH' && $message->{original_id};
+
+    my $completion_future = delete $self->pending_requests->{$message->{original_id}};
+
+    return unless $completion_future;
+
+    $completion_future->done($message);
```

Comment:
I think this can fail if there was a timeout, since you can't `->done` a ready Future

Reply:
in case of timeout we remove key with id from `$self->pending_requests`,  and this line `return unless $completion_future;` should protect us.